### PR TITLE
Refactor to share code in DecomposeComplexOps pass

### PR DESCRIPTION
**Pull request by pashu123**
Share code in `log_softmax_backward` and `softmax_backward` ops.

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -211,14 +221,12 @@ class DecomposeAtenTanhBackwardOp
     Value gradMulTanhSquare = rewriter.create<AtenMulTensorOp>(
         loc, tensorType, tanhSquare, gradOutput);
 
-    Value alpha =
-        rewriter.create<ConstantFloatOp>(loc, rewriter.getF64FloatAttr(1));
-    Value newGrad = rewriter.create<AtenSubTensorOp>(
-        loc, tensorType, gradOutput, gradMulTanhSquare, alpha);
+    Value newGrad = createAtenSubTensorOp(rewriter, loc, tensorType, gradOutput,
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:224]: nit: `createAtenSubTensorOp` seems to be too specific but strictly speak we also created ops other than just `atenSubTensorOp`. Maybe just `createTensorSub` or `createTensorSubPattern` to indicate the generated ops are for tensors subtraction? And speaking of naming, `createAtenSum` seems to be better called `createSumAlongDimension` or `createSumAlongDimensionPattern`. What do you think?

**Review by cathyzhyi — COMMENTED** [review]
Thanks for the refactoring!

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -163,22 +184,11 @@ class DecomposeAten_SoftmaxBackwardDataOp
...
-    Value temp =
-        rewriter.create<AtenMulTensorOp>(loc, tensorType, output, sumBroadcast);
-
-    // newGrad - temp
-    Value sub = createAtenSubTensorOp(rewriter, loc, tensorType, newGrad, temp);
-    rewriter.replaceOp(op, sub);
+    Value result = createSoftmaxBackwardCommonKernel(
+        rewriter, loc, op, tensorType, newGrad, output, newGrad, dim);
+    if(!result)
+        return failure();
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:190]: nit: an explicit error msg would make future debugging easier.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -211,14 +221,12 @@ class DecomposeAtenTanhBackwardOp
...
-        rewriter.create<ConstantFloatOp>(loc, rewriter.getF64FloatAttr(1));
-    Value newGrad = rewriter.create<AtenSubTensorOp>(
-        loc, tensorType, gradOutput, gradMulTanhSquare, alpha);
+    Value newGrad = createAtenSubTensorOp(rewriter, loc, tensorType, gradOutput,
+                                          gradMulTanhSquare);
     rewriter.replaceOp(op, newGrad);
     return success();
-  }
-};
+    }
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:228]: Nit: remember to clang-format

**Review by ramiro050 — COMMENTED** [review]
Thanks for this!

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 1 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:224]: Yes, the naming conventions can be improved. I can change the names as suggested in this PR.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 2 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:190]: Yup. thanks.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:228]: Oops. Thanks for catching this.

**Review by cathyzhyi — APPROVED** [review]

**Pull request merged**